test-tools/src/main/java/org/terracotta/utilities/test/rules/TestRetryer.java

```diff
@@ -62,101 +63,112 @@
 
   @SafeVarargs @SuppressWarnings("varargs") // Creating a stream from an array is safe
   public static <T> TestRetryer<T, T> tryValues(T... values) {
-    return tryValues(Stream.of(values));
+    return tryValues(asList(values));
   }
 
-  public static <T> TestRetryer<T, T> tryValues(Stream<T> values) {
-    return tryValues(values, Function.identity());
+  public static <T> TestRetryer<T, T> tryValues(Collection<? extends T> values) {
+    return new TestRetryer<>(values::stream, Function.identity(), EnumSet.noneOf(OutputIs.class));
   }
 
-  public static <T, R> TestRetryer<T, R> tryValues(Stream<T> values, Function<T, R> mapper) {
-    return tryValues(values, mapper, EnumSet.noneOf(OutputIs.class));
+  private TestRetryer(Supplier<Stream<? extends T>> values, Function<T, R> mapper, Set<OutputIs> outputIs) {
+    this.inputs = () -> values.get().map(Objects::requireNonNull);
```

Comment:
The null check described here is delayed -- not performed until the stream is actually consumed.  There can be quite a delay and, if one of the later values is null, may never observe that a value is null.  For the sake of sanity, we may be better off making a copy of the value set and checking them then.

Reply:
Fixed... added explicit null checks to the `Collection` static builder method.